README.md
# backbone-paginated-collection

[![Build Status](https://secure.travis-ci.org/jmorrell/backbone-paginated-collection.png?branch=master)](http://travis-ci.org/jmorrell/backbone-paginated-collection)

Create a read-only paginated version of a backbone collection that stays in sync.

```javascript
var superset = new Backbone.Collection(/* ... */);

// By default there are 20 models per page, but you can configure this
var paginated = new PaginatedCollection(superset, { perPage: 100 });

// Assuming superset.length === 401
assert(paginated.getNumPages() === 4);
assert(paginated.getPage() === 0);
assert(paginated.length === 100);
assert(paginated.hasNextPage());
assert(!paginated.hasPrevPage());

// Go to the next page
paginated.nextPage();
assert(paginated.getPage() === 1);

// Move to the last page
paginated.setPage(3);
assert(paginated.length === 1);
```

## Methods

### new PaginatedCollection

Initialize a new PaginatedCollection by passing in the original collection and optionally
an options hash with the number of models per page. If no `perPage` argument is passed
the collection will always maintain the length of the original collection.

```javascript
var paginated = new PaginatedCollection(originalCollection);

// or

var paginated = new PaginatedCollection(originalCollection, { perPage: 15 });
```

### paginated.setPerPage(perPage)

Change the number of models displayed per page. This will reset the current page to 0.

### paginated.setPage(page)

Change the page. If the page is less than 0, it will be set to 0. If it is longer than
the number of pages, the last page will be selected.

### paginated.getPerPage()

Return the current setting for number of models per page.

### paginated.getNumPages()

Return the current number of pages.

### paginated.getPage()

Return the current page. E.G. if this returns 0, you're on the first page.

### paginated.hasNextPage()

Returns true if this is not the last page.

### paginated.hasPrevPage()

Returns true if this is not the first page.

### paginated.movePage(delta)

Move `delta` pages forwards or backwards (if `delta` is negative).

Ex: `paginated.movePage(-2)` will move two pages back.

### paginated.nextPage()

Move to the next page. Equivalent to `paginated.movePage(1)`.

### paginated.prevPage()

Move to the previous page. Equivalent to `paginated.movePage(-1)`.

### paginated.firstPage()

Move to the first page of the collection. Equivalent to `paginated.setPage(0)`.

### paginated.lastPage()

Move to the last page of the collection. Equivalent to `paginated.setPage(paginated.getNumPages() - 1)`.

### paginated.removePagination()

Get rid of any paginated settings. This means the paginated collection
will always be equal to the superset.

### paginated.superset()

Return a reference to the original collection.

### paginated.destroy()

Remove all ties to the superset and stop updating. Will now be garbage 
collected when it falls out of scope.

## Events

`add`, `remove`, `change`, `reset` should fire as you expect.

`paginated:change:perPage` - Fired whenever the number of models per page is changed. If you
                             remove the pagination settings, `perPage` will be passed as `null`.

`paginated:change:page` - Fired whenever the page is changed.

`paginated:destroy' - Fired when the proxy is destroyed

## Installation

### Usage with Bower

Install with [Bower](http://bower.io):

```
bower install backbone-paginated-collection
```

The component can be used as a Common JS module, an AMD module, or a global.

### Usage with Browserify

Install with npm, use with [Browserify](http://browserify.org/)

```
> npm install backbone-paginated-collection
```

and in your code

```javascript
var PaginatedCollection = require('backbone-paginated-collection');
```

### Usage as browser global

You can include `backbone-paginated-collection.js` directly in a script tag. Make 
sure that it is loaded after underscore and backbone. It's exported as `PaginatedCollection`
on the global object.

```HTML
<script src="underscore.js"></script>
<script src="backbone.js"></script>
<script src="backbone-paginated-collection.js"></script>
```

## Testing

Install [Node](http://nodejs.org) (comes with npm) and Bower.

From the repo root, install the project's development dependencies:

```
npm install
bower install
```

Testing relies on the Karma test-runner. If you'd like to use Karma to
automatically watch and re-run the test file during development, it's easiest
to globally install Karma and run it from the CLI.

```
npm install -g karma
karma start
```

To run the tests in Firefox, just once, as CI would:

```
npm test
```

## License

MIT


CHANGELOG.md
v0.3.5

* Update dependency version requirements

v0.3.4

* Recalculate pagination on `sort`

v0.3.3

* Updated build with latest backbone-collection-proxy.

v0.3.2

* Added a new event when number of pages changes
* Added methods to jump to first or last page
* Fixed bug when length of superset is zero

v0.3.1

* Fixed bug when superset is empty
* Added ability to destroy proxy

v0.3.0

* Reworked the way models are added / removed
* Added ability to remove pagination
* Made commands that don't otherwise return anything chainable

v0.2.0

* Switched to browserify / grunt build process

v0.1.0

* Initial release


index.js

var _ = require('underscore');
var Backbone = require('backbone');
var proxyCollection = require('backbone-collection-proxy');

function getPageLimits() {
  var start = this.getPage() * this.getPerPage();
  var end = start + this.getPerPage();
  return [start, end];
}

function updatePagination() {
  var pages = getPageLimits.call(this);
  this._collection.reset(this.superset().slice(pages[0], pages[1]));
}

function updateNumPages() {
  var currentNumPages = this._totalPages;
  var length = this.superset().length;
  var perPage = this.getPerPage();

  // If the # of objects can be exactly divided by the number
  // of pages, it would leave an empty last page if we took
  // the floor.
  var totalPages = length % perPage === 0 ?
    (length / perPage) : Math.floor(length / perPage) + 1;

  var numPagesChanged = this._totalPages !== totalPages;
  this._totalPages = totalPages;

  if (numPagesChanged) {
    this.trigger('paginated:change:numPages', { numPages: totalPages });
  }

  // Test to see if we are past the last page, and if so,
  // move back. Return true so that we can test to see if
  // this happened.
  if (this.getPage() >= totalPages) {
    this.setPage(totalPages - 1);
    return true;
  }
}

function recalculatePagination() {
  if (updateNumPages.call(this)) { return; }
  updatePagination.call(this);
}

// Given two arrays of backbone models, with at most one model added
// and one model removed from each, return the model in arrayA that
// is not in arrayB or undefined.
function difference(arrayA, arrayB) {
  var maxLength = _.max([ arrayA.length, arrayB.length ]);

  for (var i = 0, j = 0; i < maxLength; i += 1, j += 1) {
    if (arrayA[i] !== arrayB[j]) {
      if (arrayB[i-1] === arrayA[i]) {
        j -= 1;
      } else if (arrayB[i+1] === arrayA[i]) {
        j += 1;
      } else {
        return arrayA[i];
      }
    }
  }
}

function onAddRemove(model, collection, options) {
  if (updateNumPages.call(this)) { return; }

  var pages = getPageLimits.call(this);
  var start = pages[0], end = pages[1];

  // We are only adding and removing at most one model at a time,
  // so we can find just those two models. We could probably rewrite
  // `collectionDifference` to only make on pass instead of two. This
  // is a bottleneck on the total size of collections. I was getting
  // slow unit tests around 30,000 models / page in Firefox.
  var toAdd = difference(this.superset().slice(start, end), this._collection.toArray());
  var toRemove = difference(this._collection.toArray(), this.superset().slice(start, end));

  if (toRemove) {
    this._collection.remove(toRemove);
  }

  if (toAdd) {
    this._collection.add(toAdd, {
      at: this.superset().indexOf(toAdd) - start
    });
  }
}

function Paginated(superset, options) {
  // Save a reference to the original collection
  this._superset = superset;

  // The idea is to keep an internal backbone collection with the paginated
  // set, and expose limited functionality.
  this._collection = new Backbone.Collection(superset.toArray());
  this._page = 0;
  this.setPerPage((options && options.perPage) ? options.perPage : null);

  proxyCollection(this._collection, this);

  this.listenTo(this._superset, 'add remove', onAddRemove);
  this.listenTo(this._superset, 'reset sort', recalculatePagination);
}

var methods = {

  removePagination: function() {
    this.setPerPage(null);
    return this;
  },

  setPerPage: function(perPage) {
    this._perPage = perPage;
    recalculatePagination.call(this);
    this.setPage(0);

    this.trigger('paginated:change:perPage', {
      perPage: perPage,
      numPages: this.getNumPages()
    });

    return this;
  },

  setPage: function(page) {
    // The lowest page we could set
    var lowerLimit = 0;
    // The highest page we could set
    var upperLimit = this.getNumPages() - 1;

    // If the page is higher or lower than these limits,
    // set it to the limit.
    page = page > lowerLimit ? page : lowerLimit;
    page = page < upperLimit ? page : upperLimit;
    page = page < 0 ? 0 : page;

    this._page = page;
    updatePagination.call(this);

    this.trigger('paginated:change:page', { page: page });
    return this;
  },

  getPerPage: function() {
    return this._perPage || this.superset().length || 1;
  },

  getNumPages: function() {
    return this._totalPages;
  },

  getPage: function() {
    return this._page;
  },

  hasNextPage: function() {
    return this.getPage() < this.getNumPages() - 1;
  },

  hasPrevPage: function() {
    return this.getPage() > 0;
  },

  nextPage: function() {
    this.movePage(1);
    return this;
  },

  prevPage: function() {
    this.movePage(-1);
    return this;
  },

  firstPage: function() {
    this.setPage(0);
  },

  lastPage: function() {
    this.setPage(this.getNumPages() - 1);
  },

  movePage: function(delta) {
    this.setPage(this.getPage() + delta);
    return this;
  },

  superset: function() {
    return this._superset;
  },

  destroy: function() {
    this.stopListening();
    this._collection.reset([]);
    this._superset = this._collection;
    this._page = 0;
    this._totalPages = 0;
    this.length = 0;

    this.trigger('paginated:destroy');
  }

};

// Build up the prototype
_.extend(Paginated.prototype, methods, Backbone.Events);

module.exports =  Paginated;



Gruntfile.js
module.exports = function (grunt) {
  grunt.initConfig({

    browserify: {
      basic: {
        src: [],
        dest: './backbone-paginated-collection.js',
        options: {
          external: [ 'underscore', 'backbone' ],
          alias: ['./index.js:backbone-paginated-collection']
        }
      }
    },

    umd: {
      default: {
        src: './backbone-paginated-collection.js',
        template: './templates/umd.hbs',
        objectToExport: "require('backbone-paginated-collection')",
        globalAlias: 'PaginatedCollection',
        deps: {
          'default': ['_', 'Backbone'],
          amd: ['underscore', 'backbone'],
          cjs: ['underscore', 'backbone'],
          global: ['_', 'Backbone']
        },
        browserifyMapping: '{"backbone":Backbone,"underscore":_}'
      }
    }

  });

  grunt.loadNpmTasks('grunt-browserify');
  grunt.loadNpmTasks('grunt-umd');

  grunt.registerTask('default', ['browserify', 'umd']);
};


karma.conf.js
module.exports = function(config) {
  config.set({

    // base path, that will be used to resolve files and exclude
    basePath: '',

    // list of files / patterns to load in the browser
    files: [
      'bower_components/underscore/underscore.js',
      'bower_components/backbone/backbone.js',
      'backbone-paginated-collection.js',
      'bower_components/chai/chai.js',
      'test/**/*.js'
    ],

    frameworks: ["mocha"],

    // use dots reporter, as travis terminal does not support escaping sequences
    // possible values: 'dots', 'progress', 'junit', 'teamcity'
    // CLI --reporters progress
    reporters: ['dots', 'osx'],

    // web server port
    // CLI --port 9876
    port: 9876,

    // cli runner port
    // CLI --runner-port 9100
    runnerPort: 9100,

    // enable / disable colors in the output (reporters and logs)
    // CLI --colors --no-colors
    colors: true,

    // level of logging
    // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
    // CLI --log-level debug
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    // CLI --auto-watch --no-auto-watch
    autoWatch: true,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    // CLI --browsers Chrome,Firefox,Safari
    browsers: [ process.env.TRAVIS ? 'Firefox' : 'Chrome' ],

    // If browser does not capture in given timeout [ms], kill it
    // CLI --capture-timeout 5000
    captureTimeout: 5000,

    // Auto run tests on start (when browsers are captured) and exit
    // CLI --single-run --no-single-run
    singleRun: false,

    // report which specs are slower than 500ms
    // CLI --report-slower-than 500
    reportSlowerThan: 500,

    plugins: [
      'karma-mocha',
      'karma-chrome-launcher',
      'karma-firefox-launcher',
      'karma-osx-reporter'
    ]

  });
};



backbone-paginated-collection.js
(function(root, factory) {
    if(typeof exports === 'object') {
        module.exports = factory(require('underscore'), require('backbone'));
    }
    else if(typeof define === 'function' && define.amd) {
        define(['underscore', 'backbone'], factory);
    }
    else {
        root.PaginatedCollection = factory(root._, root.Backbone);
    }
}(this, function(_, Backbone) {
var require=function(name){return {"backbone":Backbone,"underscore":_}[name];};
require=(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);throw new Error("Cannot find module '"+o+"'")}var f=n[o]={exports:{}};t[o][0].call(f.exports,function(e){var n=t[o][1][e];return s(n?n:e)},f,f.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({"PxKHrf":[function(require,module,exports){

var _ = require('underscore');
var Backbone = require('backbone');
var proxyCollection = require('backbone-collection-proxy');

function getPageLimits() {
  var start = this.getPage() * this.getPerPage();
  var end = start + this.getPerPage();
  return [start, end];
}

function updatePagination() {
  var pages = getPageLimits.call(this);
  this._collection.reset(this.superset().slice(pages[0], pages[1]));
}

function updateNumPages() {
  var currentNumPages = this._totalPages;
  var length = this.superset().length;
  var perPage = this.getPerPage();

  // If the # of objects can be exactly divided by the number
  // of pages, it would leave an empty last page if we took
  // the floor.
  var totalPages = length % perPage === 0 ?
    (length / perPage) : Math.floor(length / perPage) + 1;

  var numPagesChanged = this._totalPages !== totalPages;
  this._totalPages = totalPages;

  if (numPagesChanged) {
    this.trigger('paginated:change:numPages', { numPages: totalPages });
  }

  // Test to see if we are past the last page, and if so,
  // move back. Return true so that we can test to see if
  // this happened.
  if (this.getPage() >= totalPages) {
    this.setPage(totalPages - 1);
    return true;
  }
}

function recalculatePagination() {
  if (updateNumPages.call(this)) { return; }
  updatePagination.call(this);
}

// Given two arrays of backbone models, with at most one model added
// and one model removed from each, return the model in arrayA that
// is not in arrayB or undefined.
function difference(arrayA, arrayB) {
  var maxLength = _.max([ arrayA.length, arrayB.length ]);

  for (var i = 0, j = 0; i < maxLength; i += 1, j += 1) {
    if (arrayA[i] !== arrayB[j]) {
      if (arrayB[i-1] === arrayA[i]) {
        j -= 1;
      } else if (arrayB[i+1] === arrayA[i]) {
        j += 1;
      } else {
        return arrayA[i];
      }
    }
  }
}

function onAddRemove(model, collection, options) {
  if (updateNumPages.call(this)) { return; }

  var pages = getPageLimits.call(this);
  var start = pages[0], end = pages[1];

  // We are only adding and removing at most one model at a time,
  // so we can find just those two models. We could probably rewrite
  // `collectionDifference` to only make on pass instead of two. This
  // is a bottleneck on the total size of collections. I was getting
  // slow unit tests around 30,000 models / page in Firefox.
  var toAdd = difference(this.superset().slice(start, end), this._collection.toArray());
  var toRemove = difference(this._collection.toArray(), this.superset().slice(start, end));

  if (toRemove) {
    this._collection.remove(toRemove);
  }

  if (toAdd) {
    this._collection.add(toAdd, {
      at: this.superset().indexOf(toAdd) - start
    });
  }
}

function Paginated(superset, options) {
  // Save a reference to the original collection
  this._superset = superset;

  // The idea is to keep an internal backbone collection with the paginated
  // set, and expose limited functionality.
  this._collection = new Backbone.Collection(superset.toArray());
  this._page = 0;
  this.setPerPage((options && options.perPage) ? options.perPage : null);

  proxyCollection(this._collection, this);

  this.listenTo(this._superset, 'add remove', onAddRemove);
  this.listenTo(this._superset, 'reset sort', recalculatePagination);
}

var methods = {

  removePagination: function() {
    this.setPerPage(null);
    return this;
  },

  setPerPage: function(perPage) {
    this._perPage = perPage;
    recalculatePagination.call(this);
    this.setPage(0);

    this.trigger('paginated:change:perPage', {
      perPage: perPage,
      numPages: this.getNumPages()
    });

    return this;
  },

  setPage: function(page) {
    // The lowest page we could set
    var lowerLimit = 0;
    // The highest page we could set
    var upperLimit = this.getNumPages() - 1;

    // If the page is higher or lower than these limits,
    // set it to the limit.
    page = page > lowerLimit ? page : lowerLimit;
    page = page < upperLimit ? page : upperLimit;
    page = page < 0 ? 0 : page;

    this._page = page;
    updatePagination.call(this);

    this.trigger('paginated:change:page', { page: page });
    return this;
  },

  getPerPage: function() {
    return this._perPage || this.superset().length || 1;
  },

  getNumPages: function() {
    return this._totalPages;
  },

  getPage: function() {
    return this._page;
  },

  hasNextPage: function() {
    return this.getPage() < this.getNumPages() - 1;
  },

  hasPrevPage: function() {
    return this.getPage() > 0;
  },

  nextPage: function() {
    this.movePage(1);
    return this;
  },

  prevPage: function() {
    this.movePage(-1);
    return this;
  },

  firstPage: function() {
    this.setPage(0);
  },

  lastPage: function() {
    this.setPage(this.getNumPages() - 1);
  },

  movePage: function(delta) {
    this.setPage(this.getPage() + delta);
    return this;
  },

  superset: function() {
    return this._superset;
  },

  destroy: function() {
    this.stopListening();
    this._collection.reset([]);
    this._superset = this._collection;
    this._page = 0;
    this._totalPages = 0;
    this.length = 0;

    this.trigger('paginated:destroy');
  }

};

// Build up the prototype
_.extend(Paginated.prototype, methods, Backbone.Events);

module.exports =  Paginated;


},{"backbone":false,"backbone-collection-proxy":3,"underscore":false}],"backbone-paginated-collection":[function(require,module,exports){
module.exports=require('PxKHrf');
},{}],3:[function(require,module,exports){

var _ = require('underscore');
var Backbone = require('backbone');

// Methods in the collection prototype that we won't expose
var blacklistedMethods = [
  "_onModelEvent", "_prepareModel", "_removeReference", "_reset", "add",
  "initialize", "sync", "remove", "reset", "set", "push", "pop", "unshift",
  "shift", "sort", "parse", "fetch", "create", "model", "off", "on",
  "listenTo", "listenToOnce", "bind", "trigger", "once", "stopListening"
];

var eventWhiteList = [
  'add', 'remove', 'reset', 'sort', 'destroy'
];

function proxyCollection(from, target) {

  function updateLength() {
    target.length = from.length;
  }

  function pipeEvents(eventName) {
    var args = _.toArray(arguments);
    var isChangeEvent = eventName === 'change' ||
                        eventName.slice(0, 7) === 'change:';

    // In the case of a `reset` event, the Collection.models reference
    // is updated to a new array, so we need to update our reference.
    if (eventName === 'reset') {
      target.models = from.models;
    }

    if (_.contains(eventWhiteList, eventName)) {
      if (_.contains(['add', 'remove', 'destory'], eventName)) {
        args[2] = target;
      } else if (_.contains(['reset', 'sort'], eventName)) {
        args[1] = target;
      }
      target.trigger.apply(this, args);
    } else if (isChangeEvent) {
      // In some cases I was seeing change events fired after the model
      // had already been removed from the collection.
      if (target.contains(args[1])) {
        target.trigger.apply(this, args);
      }
    }
  }

  var methods = {};

  _.each(_.functions(Backbone.Collection.prototype), function(method) {
    if (!_.contains(blacklistedMethods, method)) {
      methods[method] = function() {
        return from[method].apply(from, arguments);
      };
    }
  });

  _.extend(target, Backbone.Events, methods);

  target.listenTo(from, 'all', updateLength);
  target.listenTo(from, 'all', pipeEvents);
  target.models = from.models;

  updateLength();
  return target;
}

module.exports = proxyCollection;


},{"backbone":false,"underscore":false}]},{},[])
;
return require('backbone-paginated-collection');

}));
